File: others/js/obf.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

class JSObfuscator {
    constructor() {
        this.config = {
            renameVariables: true,
            encryptStrings: true,
            addDeadCode: true,
            flattenControl: true,
            selfDefending: true,
            debugProtection: true,
            reorderFunctions: true,
            moveLiterals: true
        };

        this.varCounter = 0;
        this.reservedWords = [
            'abstract', 'arguments', 'await', 'boolean', 'break', 'byte', 'case', 'catch', 'char',
            'class', 'const', 'continue', 'debugger', 'default', 'delete', 'do', 'double', 'else',
            'enum', 'eval', 'export', 'extends', 'false', 'final', 'finally', 'float', 'for',
            'function', 'goto', 'if', 'implements', 'import', 'in', 'instanceof', 'int', 'interface',
            'let', 'long', 'native', 'new', 'null', 'package', 'private', 'protected', 'public',
            'return', 'short', 'static', 'super', 'switch', 'synchronized', 'this', 'throw', 'throws',
            'transient', 'true', 'try', 'typeof', 'var', 'void', 'volatile', 'while', 'with', 'yield',
            'Array', 'Date', 'eval', 'function', 'hasOwnProperty', 'Infinity', 'isFinite', 'isNaN',
            'length', 'Math', 'NaN', 'name', 'Number', 'Object', 'prototype', 'String', 'undefined'
        ];

        this.identifiers = [];
        this.strings = [];
    }

    generateVarName() {
        this.varCounter++;
        return '_' + this.varCounter.toString(36);
    }

    removeComments(js) {
        return js
            .replace(/\/\/.*$/gm, '')
            .replace(/\/\*[\s\S]*?\*\//g, '');
    }

    extractIdentifiers(js) {
        const varMatches = js.match(/(?:var|let|const)\s+([a-zA-Z_$][a-zA-Z0-9_$]*)/g) || [];
        const funcMatches = js.match(/function\s+([a-zA-Z_$][a-zA-Z0-9_$]*)/g) || [];

        const varNames = varMatches.map(match => match.replace(/(?:var|let|const)\s+/, ''));
        const funcNames = funcMatches.map(match => match.replace(/function\s+/, ''));

        this.identifiers = [...new Set([...varNames, ...funcNames])];
        this.identifiers = this.identifiers.filter(id => !this.reservedWords.includes(id));
    }

    extractStrings(js) {
        const singleQuotes = js.match(/'((?:\\.|[^\\'])*)'/g) || [];
        const doubleQuotes = js.match(/"((?:\\.|[^\\"])*)"/g) || [];

        this.strings = [...singleQuotes, ...doubleQuotes];
    }

    renameIdentifiers(js) {
        if (!this.config.renameVariables) {
            return js;
        }

        for (const id of this.identifiers) {
            if (id.length <= 2 || this.reservedWords.includes(id)) {
                continue;
            }

            const newName = this.generateVarName();
            const regex = new RegExp(`\\b${id}\\b(?!\\s*:)`, 'g');
            js = js.replace(regex, newName);
        }

        return js;
    }

    encryptStrings(js) {
        if (!this.config.encryptStrings) {
            return js;
        }

        const decoder = this.generateVarName();
        const key = Math.floor(Math.random() * 50) + 1;

        const decoderFunc = `function ${decoder}(s){var r='',i=0;for(;i<s.length;i++)r+=String.fromCharCode(s.charCodeAt(i)^${key});return r}`;

        for (const string of this.strings) {
            if (string.length <= 3) continue;

            const content = string.substring(1, string.length - 1);
            if (content.length <= 2) continue;

            let encrypted = '';
            for (let i = 0; i < content.length; i++) {
                const charCode = content.charCodeAt(i) ^ key;
                encrypted += '\\x' + charCode.toString(16).padStart(2, '0');
            }

            const quote = string[0];
            js = js.replace(string, `${decoder}(${quote}${encrypted}${quote})`);
        }

        return decoderFunc + ';' + js;
    }

    moveLiterals(js) {
        if (!this.config.moveLiterals) {
            return js;
        }

        const numMatches = js.match(/\b(\d+)\b/g) || [];
        const nums = [...new Set(numMatches)];
        const numVars = {};

        for (const num of nums) {
            if (parseInt(num) < 10) continue;

            const varName = this.generateVarName();
            numVars[num] = varName;
        }

        let declarations = '';
        for (const [num, varName] of Object.entries(numVars)) {
            declarations += `var ${varName}=${num};`;
            const regex = new RegExp(`\\b${num}\\b`, 'g');
            js = js.replace(regex, varName);
        }

        return declarations + js;
    }

    flattenControlFlow(js) {
        if (!this.config.flattenControl) {
            return js;
        }

        const ifElseRegex = /if\s*\((.*?)\)\s*{(.*?)}\s*else\s*{(.*?)}/gs;
        let match;
        let result = js;

        while ((match = ifElseRegex.exec(js)) !== null) {
            const condition = match[1];
            const ifBlock = match[2];
            const elseBlock = match[3];

            const switchVar = this.generateVarName();
            const case1 = Math.floor(Math.random() * 100) + 1;
            const case2 = Math.floor(Math.random() * 100) + 101;

            const replacement = `var ${switchVar}=${condition}?${case1}:${case2};switch(${switchVar}){case ${case1}:${ifBlock}break;case ${case2}:${elseBlock}break;}`;
            result = result.replace(match[0], replacement);
        }

        return result;
    }

    reorderFunctions(js) {
        if (!this.config.reorderFunctions) {
            return js;
        }

        const funcRegex = /function\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*\([^)]*\)\s*{(?:[^{}]*|(?:(?!{).)*{(?:[^{}]*|(?:(?!{).)*{[^{}]*})*})*}/g;
        const functions = [];
        let match;

        while ((match = funcRegex.exec(js)) !== null) {
            functions.push(match[0]);
        }

        if (functions.length <= 1) {
            return js;
        }

        let result = js;
        for (const func of functions) {
            result = result.replace(func, '/* FUNCTION_PLACEHOLDER */');
        }

        functions.sort(() => Math.random() - 0.5);

        for (const func of functions) {
            result = result.replace('/* FUNCTION_PLACEHOLDER */', func);
        }

        return result;
    }

    addDeadCode(js) {
        if (!this.config.addDeadCode) {
            return js;
        }

        const deadCodeSamples = [
            "if(false){console.log('Dead code');}",
            "var __d=Date.now();if(__d<0){throw new Error('Time travel detected');}",
            "try{if(typeof window==='undefined'&&typeof window.__z==='string'){}}catch(e){}",
            "function __zzz(){return Math.random()<0?true:false}if(__zzz()){}",
            "var __counter=0;while(__counter<0){__counter++;console.log('Never runs');}"
        ];

        const positions = [];
        const semicolons = /[;}]/g;
        let match;

        while ((match = semicolons.exec(js)) !== null) {
            positions.push(match.index);
        }

        if (positions.length === 0) {
            return js;
        }

        let result = js;
        let offset = 0;
        const insertCount = Math.min(3, positions.length);

        for (let i = 0; i < insertCount; i++) {
            const randomIndex = Math.floor(Math.random() * positions.length);
            const position = positions[randomIndex];
            positions.splice(randomIndex, 1);

            const deadCode = deadCodeSamples[Math.floor(Math.random() * deadCodeSamples.length)];
            result = result.slice(0, position + offset + 1) + deadCode + result.slice(position + offset + 1);

            offset += deadCode.length;
        }

        return result;
    }

    compactCode(js) {
        return js
            .replace(/\s+/g, ' ')
            .replace(/\s*([{}:;,=+\-*\/&|<>!?()[\]])\s*/g, '$1')
            .replace(/\s*\n\s*/g, '')
            .replace(/;}/g, '}');
    }

    addSelfDefense(js) {
        if (!this.config.selfDefending) {
            return js;
        }

        const checkFuncName = this.generateVarName();
        const originalCode = this.generateVarName();
        const interval = this.generateVarName();

        const codeHash = crypto.createHash('md5').update(js).digest('hex');

        const selfDefense = `
(function(){
    var ${originalCode}='${codeHash}';
    function ${checkFuncName}(){
        var currentCode=document.currentScript.text.replace(/\\s+/g,'');
        var hash='';
        for(var i=0;i<currentCode.length;i++){
            hash=((hash<<5)-hash)+currentCode.charCodeAt(i);
            hash=hash&hash;
        }
        hash=hash.toString(16);
        if(hash!==${originalCode}){
            document.body.innerHTML='';
            window.location.href='about:blank';
        }
    }
    var ${interval}=setInterval(${checkFuncName},2000);
})();`;

        return selfDefense + js;
    }

    addDebugProtection(js) {
        if (!this.config.debugProtection) {
            return js;
        }

        const debugProtection = `
(function(){
    setInterval(function(){
        debugger;
    },100);

    var start=new Date();
    debugger;
    var end=new Date();

    if(end-start>100){
        document.body.innerHTML='';
        window.location.href='about:blank';
    }
})();`;

        return debugProtection + js;
    }

    obfuscate(js) {
        js = this.removeComments(js);

        this.extractIdentifiers(js);
        this.extractStrings(js);

        if (this.config.renameVariables) {
            js = this.renameIdentifiers(js);
        }

        if (this.config.encryptStrings) {
            js = this.encryptStrings(js);
        }

        if (this.config.moveLiterals) {
            js = this.moveLiterals(js);
        }

        if (this.config.flattenControl) {
            js = this.flattenControlFlow(js);
        }

        if (this.config.reorderFunctions) {
            js = this.reorderFunctions(js);
        }

        if (this.config.addDeadCode) {
            js = this.addDeadCode(js);
        }

        js = this.compactCode(js);

        if (this.config.selfDefending) {
            js = this.addSelfDefense(js);
        }

        if (this.config.debugProtection) {
            js = this.addDebugProtection(js);
        }

        return js;
    }
}

function main() {
    const args = process.argv.slice(2);

    if (args.length < 1) {
        console.log(`Usage: ${process.argv[1]} input.js [output.js]`);
        process.exit(1);
    }

    const inputFile = args[0];
    const outputFile = args[1] || path.basename(inputFile, path.extname(inputFile)) + '.obf.js';

    if (!fs.existsSync(inputFile)) {
        console.error(`Error: Input file '${inputFile}' not found`);
        process.exit(1);
    }

    try {
        const jsCode = fs.readFileSync(inputFile, 'utf8');

        const obfuscator = new JSObfuscator();
        const obfuscated = obfuscator.obfuscate(jsCode);

        fs.writeFileSync(outputFile, obfuscated, 'utf8');
        console.log(`Obfuscation complete. Output saved to '${outputFile}'`);
    } catch (err) {
        console.error(`Error: ${err.message}`);
        process.exit(1);
    }
}

main();

File: others/python/obf.py
#!/usr/bin/env python3
import sys
import re
import random
import hashlib
import base64
import os

class JSObfuscator:
    def __init__(self):
        self.config = {
            'rename_vars': True,
            'encrypt_strings': True,
            'add_dead_code': True,
            'flatten_control': True,
            'self_defend': True,
            'debug_protect': True,
            'reorder_funcs': True,
            'move_literals': True
        }

        self.var_counter = 0
        self.reserved_words = [
            'abstract', 'arguments', 'await', 'boolean', 'break', 'byte', 'case', 'catch', 'char',
            'class', 'const', 'continue', 'debugger', 'default', 'delete', 'do', 'double', 'else',
            'enum', 'eval', 'export', 'extends', 'false', 'final', 'finally', 'float', 'for',
            'function', 'goto', 'if', 'implements', 'import', 'in', 'instanceof', 'int', 'interface',
            'let', 'long', 'native', 'new', 'null', 'package', 'private', 'protected', 'public',
            'return', 'short', 'static', 'super', 'switch', 'synchronized', 'this', 'throw', 'throws',
            'transient', 'true', 'try', 'typeof', 'var', 'void', 'volatile', 'while', 'with', 'yield'
        ]
        self.identifiers = []
        self.strings = []

    def gen_var_name(self):
        self.var_counter += 1
        return '_' + base64.b32encode(str(self.var_counter).encode()).decode().replace('=', '').lower()[:5]

    def remove_comments(self, js):
        js = re.sub(r'\/\/.*$', '', js, flags=re.MULTILINE)
        js = re.sub(r'\/\*[\s\S]*?\*\/', '', js)
        return js

    def extract_identifiers(self, js):
        var_matches = re.findall(r'(?:var|let|const)\s+([a-zA-Z_$][a-zA-Z0-9_$]*)', js)
        func_matches = re.findall(r'function\s+([a-zA-Z_$][a-zA-Z0-9_$]*)', js)

        self.identifiers = list(set(var_matches + func_matches))
        self.identifiers = [id for id in self.identifiers if id not in self.reserved_words]

    def extract_strings(self, js):
        single_quotes = re.findall(r'\'((?:\\\\.|[^\\\\\'])*)\'', js)
        double_quotes = re.findall(r'"((?:\\\\.|[^\\\\"])*)"', js)

        self.strings = single_quotes + double_quotes

    def rename_identifiers(self, js):
        for id in self.identifiers:
            if len(id) <= 2 or id in self.reserved_words:
                continue

            new_name = self.gen_var_name()
            js = re.sub(r'\b' + re.escape(id) + r'\b(?!\s*:)', new_name, js)

        return js

    def encrypt_strings(self, js):
        if not self.config['encrypt_strings']:
            return js

        decoder = self.gen_var_name()
        key = random.randint(1, 50)

        decoder_func = f"function {decoder}(s){{var r='',i=0;for(;i<s.length;i++)r+=String.fromCharCode(s.charCodeAt(i)^{key});return r}}"

        for string in self.strings:
            if len(string) <= 2:
                continue

            encrypted = ''
            for char in string:
                encrypted += '\\x' + format(ord(char) ^ key, '02x')

            js = js.replace(f"'{string}'", f"{decoder}('{encrypted}')")
            js = js.replace(f'"{string}"', f'{decoder}("{encrypted}")')

        return decoder_func + ";" + js

    def move_literals(self, js):
        if not self.config['move_literals']:
            return js

        num_matches = re.findall(r'\b(\d+)\b', js)
        nums = list(set(num_matches))
        num_vars = {}

        for num in nums:
            if int(num) < 10:
                continue

            var_name = self.gen_var_name()
            num_vars[num] = var_name

        declarations = ''
        for num, var in num_vars.items():
            declarations += f"var {var}={num};"
            js = re.sub(r'\b' + num + r'\b', var, js)

        return declarations + js

    def flatten_control_flow(self, js):
        if not self.config['flatten_control']:
            return js

        patterns = re.finditer(r'if\s*\((.*?)\)\s*{(.*?)}\s*else\s*{(.*?)}', js, re.DOTALL)

        for match in patterns:
            condition = match.group(1)
            if_block = match.group(2)
            else_block = match.group(3)

            switch_var = self.gen_var_name()
            case1 = random.randint(1, 100)
            case2 = random.randint(101, 200)

            replacement = f"var {switch_var}={condition}?{case1}:{case2};switch({switch_var}){{case {case1}:{if_block}break;case {case2}:{else_block}break;}}"
            js = js.replace(match.group(0), replacement)

        return js

    def reorder_functions(self, js):
        if not self.config['reorder_funcs']:
            return js

        func_matches = re.finditer(r'function\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*\([^)]*\)\s*{(?:[^{}]*|(?R))*}', js, re.DOTALL)
        functions = [match.group(0) for match in func_matches]

        if len(functions) <= 1:
            return js

        for func in functions:
            js = js.replace(func, '/* FUNCTION_PLACEHOLDER */', 1)

        random.shuffle(functions)

        for func in functions:
            js = js.replace('/* FUNCTION_PLACEHOLDER */', func, 1)

        return js

    def add_dead_code(self, js):
        if not self.config['add_dead_code']:
            return js

        dead_code = [
            "if(false){console.log('x');}",
            "var __d=Date.now();if(__d<0){}",
            "try{if(typeof window==='undefined'&&window.__z==='x'){}}catch(e){}",
            "function __x(){return Math.random()<0}if(__x()){}",
            "var __c=0;while(__c<0){__c++;}"
        ]

        positions = [m.start() for m in re.finditer(r'[;}]', js)]

        if not positions:
            return js

        result = js
        inserted = 0
        random.shuffle(positions)

        for pos in positions[:3]:
            code = random.choice(dead_code)
            result = result[:pos+1] + code + result[pos+1:]
            inserted += 1

        return result

    def compact_code(self, js):
        js = re.sub(r'\s+', ' ', js)
        js = re.sub(r'\s*([{}:;,=+\-*\/&|<>!?()[\]])\s*', r'\1', js)
        js = re.sub(r';}', '}', js)
        return js

    def add_self_defense(self, js):
        if not self.config['self_defend']:
            return js

        check_func = self.gen_var_name()
        original_code = self.gen_var_name()
        interval = self.gen_var_name()

        code_hash = hashlib.md5(js.encode()).hexdigest()

        defense = f"""
(function(){{
var {original_code}='{code_hash}';
function {check_func}(){{
var c=document.currentScript.text.replace(/\\s+/g,'');
var h='';
for(var i=0;i<c.length;i++){{
h=((h<<5)-h)+c.charCodeAt(i);
h=h&h;
}}
h=h.toString(16);
if(h!=={original_code}){{
document.body.innerHTML='';
window.location='about:blank';
}}
}}
var {interval}=setInterval({check_func},2000);
}})();
"""
        return defense + js

    def add_debug_protection(self, js):
        if not self.config['debug_protect']:
            return js

        protection = """
(function(){
setInterval(function(){
debugger;
},100);
var s=new Date();
debugger;
var e=new Date();
if(e-s>100){
document.body.innerHTML='';
window.location='about:blank';
}
})();
"""
        return protection + js

    def obfuscate(self, js):
        js = self.remove_comments(js)

        self.extract_identifiers(js)
        self.extract_strings(js)

        if self.config['rename_vars']:
            js = self.rename_identifiers(js)

        if self.config['encrypt_strings']:
            js = self.encrypt_strings(js)

        if self.config['move_literals']:
            js = self.move_literals(js)

        if self.config['flatten_control']:
            js = self.flatten_control_flow(js)

        if self.config['reorder_funcs']:
            js = self.reorder_functions(js)

        if self.config['add_dead_code']:
            js = self.add_dead_code(js)

        js = self.compact_code(js)

        if self.config['self_defend']:
            js = self.add_self_defense(js)

        if self.config['debug_protect']:
            js = self.add_debug_protection(js)

        return js

def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} input.js [output.js]")
        sys.exit(1)

    input_file = sys.argv[1]

    if len(sys.argv) >= 3:
        output_file = sys.argv[2]
    else:
        base_name = os.path.splitext(input_file)[0]
        output_file = f"{base_name}.obf.js"

    if not os.path.exists(input_file):
        print(f"Error: Input file '{input_file}' not found")
        sys.exit(1)

    try:
        with open(input_file, 'r', encoding='utf-8') as f:
            js_code = f.read()

        obfuscator = JSObfuscator()
        obfuscated = obfuscator.obfuscate(js_code)

        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(obfuscated)

        print(f"Obfuscation complete. Output saved to '{output_file}'")

    except Exception as e:
        print(f"Error: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main()